crates/matrix-sdk-base/src/store/ambiguity_map.rs

```diff
@@ -154,60 +154,76 @@ impl AmbiguityCache {
         self.changes.entry(room_id.to_owned()).or_default().insert(event_id, change);
     }
 
-    async fn get(
-        &mut self,
+    async fn get_old_display_name(
+        &self,
         changes: &StateChanges,
         room_id: &RoomId,
-        member_event: &SyncRoomMemberEvent,
-    ) -> Result<(Option<DisplayNameUsers>, Option<DisplayNameUsers>)> {
+        new_event: &SyncRoomMemberEvent,
+    ) -> Result<Option<String>> {
         use MembershipState::*;
 
         let old_event = if let Some(m) = changes.state.get(room_id).and_then(|events| {
-            events.get(&StateEventType::RoomMember)?.get(member_event.state_key().as_str())
+            events.get(&StateEventType::RoomMember)?.get(new_event.state_key().as_str())
         }) {
             Some(RawMemberEvent::Sync(m.clone().cast()))
         } else {
-            self.store.get_member_event(room_id, member_event.state_key()).await?
+            self.store.get_member_event(room_id, new_event.state_key()).await?
         };
 
-        // FIXME: Use let chains once stable
-        let old_display_name = if let Some(Ok(event)) = old_event.map(|r| r.deserialize()) {
-            if matches!(event.membership(), Join | Invite) {
-                let display_name = if let Some(d) = changes.profiles.get(room_id).and_then(|p| {
-                    p.get(member_event.state_key())?.as_original()?.content.displayname.as_deref()
-                }) {
-                    Some(d.to_owned())
-                } else if let Some(d) = self
-                    .store
-                    .get_profile(room_id, member_event.state_key())
-                    .await?
-                    .and_then(|p| p.into_original()?.content.displayname)
-                {
-                    Some(d)
-                } else {
-                    event.original_content().and_then(|c| c.displayname.clone())
-                };
-
-                Some(display_name.unwrap_or_else(|| event.user_id().localpart().to_owned()))
+        let Some(Ok(old_event)) = old_event.map(|r| r.deserialize()) else { return Ok(None) };
+
+        if matches!(old_event.membership(), Join | Invite) {
```

Comment:
Should we include Knock here too?

Reply:
Yeah I think so, though that's a feature, not a refactor: 4532a4db6588cf212ff0108caed36415923728f1.